File: solutions/0023-merge-k-sorted-lists/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool operator()(ListNode* a, ListNode* b) const {
        // return true if a should come after b, i.e. a.val > b.val
        if(a and b)
        return a->val > b->val;
        return true;
    }
    ListNode* merge(ListNode* l1,ListNode* l2){
        ListNode* temp = new ListNode();
        ListNode* dummy = temp;
        while(l1 and l2){
            if(l1->val<l2->val){
                dummy->next = l1;
                l1 = l1->next;
            }
            else{
                dummy->next = l2;
                l2 = l2->next;
            }
            dummy=dummy->next;
        }
        if(l1){
            dummy->next = l1;
        }
        if(l2){
            dummy->next = l2;
        }
        return temp->next;
    }
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(lists.size()==0){
            return nullptr;
        }
        priority_queue<ListNode*,vector<ListNode*>,Solution> heap;
        for(int i=0;i<lists.size();i++){
            heap.push(lists[i]);
        }
        while(heap.size()>1){
            ListNode* a = heap.top();heap.pop();
            ListNode* b = heap.top();heap.pop();
            heap.push(merge(a,b));
        }
        return heap.top();
    }
};

File: solutions/0023-merge-k-sorted-lists/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
import heapq

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        # i will create a heap, which have the size of lists=k, we keep first smallest nodes of each lists in that heap
        # we process the smallest value of node, we pop it, and push next node of that popped node in heap
        # we keep merging the nodes and after anothers by poppin from heap
        # we do this until heap become empty
        dummy = ListNode()
        curr = dummy
        minheap = []
        # put first nodes in min heap -> (node value, node) 
        idx = 0
        for l in lists:
            if l:
                heapq.heappush(minheap,(l.val,idx,l))
                idx+=1
        # start poppin until minheap become empty

        while len(minheap)>0:
            # take the smallest one and pop
            item = heapq.heappop(minheap)
            curr.next = item[2]
            curr = curr.next
            # put the next node in heap, if exist
            if curr.next:
                heapq.heappush(minheap,(curr.next.val,idx,curr.next))
                idx+=1
        return dummy.next



        



        
